routing/probability_bimodal.go

```diff
@@ -0,0 +1,238 @@
+}
+
+// getPairProbability estimates the probability of successfully traversing to
+// toNode based on historical payment outcomes for the from node. Those outcomes
+// are passed in via the results parameter.
+func (p *probabilityBimodal) getPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity btcutil.Amount) float64 {
+
+	// We first compute the probability for the desired hop taking into
+	// account previous knowledge.
+	directProbability := p.getLiquidityProbability(
+		now, results, toNode, amt, lnwire.NewMSatFromSatoshis(capacity),
+	)
+
+	// The final probability is computed by taking into account other
+	// channels of the node.
+	return p.calculateProbability(directProbability, now, results, toNode)
+}
+
+// getLocalPairProbability computes the probability to reach toNode given a set
+// of previous learnings.
+func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex) float64 {
+
+	return 1.0
+}
+
+// getLiquidityProbability computes the probability to reach a node based
+// on the liquidity distribution in the LN.
+func (p *probabilityBimodal) getLiquidityProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity lnwire.MilliSatoshi) float64 {
+
+	// Set success and failure amounts.
+
+	// We know that we can send zero satoshis.
+	successAmount := lnwire.MilliSatoshi(0)
+	// We know that we cannot send the full capacity.
+	failAmount := capacity
+
+	result, ok := results[toNode]
+	if ok {
+		if !result.FailTime.IsZero() {
+			// Apply time decay for the amount we can not send.
+			failAmount = cannotSend(
+				result.FailAmt,
+				capacity,
+				now,
+				result.FailTime,
+				p.PenaltyHalfLife,
+			)
+		}
+		if !result.SuccessTime.IsZero() {
+			// Apply time decay for the amount we can send.
+			successAmount = canSend(
+				result.SuccessAmt,
+				now,
+				result.SuccessTime,
+				p.PenaltyHalfLife,
+			)
+		}
+	}
+
+	probability := p.probabilityFormula(
+		float64(capacity),
+		float64(successAmount),
+		float64(failAmount),
+		float64(amt),
+	)
+
+	return probability
+}
+
+// calculateProbability computes the total hop probability based on the channel
+// probability and historic forwarding data of other channels of the node.
+func (p *probabilityBimodal) calculateProbability(directProbability float64,
+	now time.Time, results NodeResults, toNode route.Vertex) float64 {
+
+	// w is a parameter which determines how strongly the other channels of
+	// a node should be incorportated, the higher the stronger.
+	w := p.AprioriWeight
+
+	// hl determines the timeliness of the previous successes/failures
+	// to be taken into account.
+	hl := p.PenaltyHalfLife
+
+	// The channel probability is weighted fully, all other results are
+	// weighted according to how recent the information is.
+	totalProbabilities := directProbability
+	totalWeights := 1.
```

Comment:
Because the result of this function is used directly and no further mixing with the direct probability is done, it is quite different compared to how it was. Especially with large amounts of channels, it seems that the direct probability hardly has any effect anymore of the final probability estimate.

Reply:
> Especially with large amounts of channels, it seems that the direct probability hardly has any effect anymore of the final probability estimate.

I really thought hard about this aspect, it may as well be a feature. I added some extended calculations to the docstring to document behavior. I also think that this approach is not too far from the current scheme, the node weight chosen is quite conservative, I think. We can set influence of this to zero with the node weight. This way one could independently test the direct probability against the current approach.